diff --git a/include/android/keycodes.h b/include/android/keycodes.h
index 282e729..17b744f 100644
--- a/include/android/keycodes.h
+++ b/include/android/keycodes.h
@@ -263,6 +263,16 @@ enum {
     AKEYCODE_RO              = 217,
     AKEYCODE_KANA            = 218,
     AKEYCODE_ASSIST          = 219,
+    AKEYCODE_WIRELESS        = 220,
+    AKEYCODE_BLUETOOTH       = 221,
+    AKEYCODE_TOUCHPAD        = 222,
+    AKEYCODE_BRIGHTNESS_UP   = 223,
+    AKEYCODE_BRIGHTNESS_DOWN = 224,
+    AKEYCODE_BRIGHTNESS_AUTO = 225,
+    AKEYCODE_CAPTURE         = 226,
+    AKEYCODE_EUROPE_1        = 227,
+    AKEYCODE_EUROPE_2        = 228,
+    AKEYCODE_SLEEP           = 229,
 
     // NOTE: If you add a new keycode here you must also add it to several other files.
     //       Refer to frameworks/base/core/java/android/view/KeyEvent.java for the full list.
