diff --git a/media/libmedia/AudioSystem.cpp b/media/libmedia/AudioSystem.cpp
index 4c41ba5..f2d484e 100644
--- a/media/libmedia/AudioSystem.cpp
+++ b/media/libmedia/AudioSystem.cpp
@@ -596,6 +596,13 @@ audio_io_handle_t AudioSystem::getOutput(audio_stream_type_t stream,
     return aps->getOutput(stream, samplingRate, format, channels, flags);
 }
 
+// Legacy constructor for Tegra OMX
+extern "C" audio_io_handle_t _ZN7android11AudioSystem9getOutputE19audio_stream_type_tjjj27audio_policy_output_flags_t(
+        audio_stream_type_t stream, uint32_t samplingRate, uint32_t format, uint32_t channels, audio_output_flags_t flags)
+{
+    return AudioSystem::getOutput(stream, samplingRate, (audio_format_t)format, channels, flags);
+}
+
 status_t AudioSystem::startOutput(audio_io_handle_t output,
                                   audio_stream_type_t stream,
                                   int session)
